src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -3057,6 +3070,32 @@ void CallData::InjectRecvInitialMetadataReadyForConfigSelectorCommitCallback(
       &recv_initial_metadata_ready_;
 }
 
+void CallData::RecvTrailingMetadataReadyForAdditionalErrorContext(
+    void* arg, grpc_error* error) {
+  auto* self = static_cast<CallData*>(arg);
+  GRPC_ERROR_REF(error);
+  if (error != GRPC_ERROR_NONE) {
+    if (self->dynamic_call_ == nullptr) {
+      error = grpc_error_set_int(
+          error, GRPC_ERROR_INT_OCCURRED_WHILE_AWAITING_NAME_RESOLUTION, true);
```

Comment:
Instead of adding a new error attribute, I suggest just adding text to the error message itself.  That way, it will actually be visible to users.

Reply:
I was going to comment this in https://github.com/grpc/grpc/pull/25158, but I don't think it's necessary to pack everything into the error message string, i.e. I don't think it's necessary to avoid reliance on `debug_error_string()`.

I know for sure that Ruby, Python, and C# all include the full JSON error string from https://github.com/grpc/grpc/blob/9eae29f8e988f7eb15989410e5cb2745a059a895/include/grpc/impl/codegen/grpc_types.h#L676 (the equivalent of `debug_error_string()` in their RPC exceptions.

I'm not sure about PHP and Node, but we could probably modify them to include it in their RPC exceptions/errors they don't already.

For C++ users, I know that at least the application code which initially motivated this actually logs `debug_error_string()` along with the RPC error status, whenever an RPC fails. So, I believe at least some number of C++ users  aware that the more verbose but detailed error message lies within `debug_error_string()`.

Overall, I'm on the side of encouraging use of `debug_error_string()`, rather than putting more and more details into the error message string (which can remain as a brief description of the error).

Reply:
One problem I have with `debug_error_string()` is that it's basically exposing internal implementation details that may change.  Note that the current format of this field is based on the existing implementation of `grpc_error`.  However, that type actually causes us a lot of problems, and I would like to see us completely eliminate that type in favor of `absl::Status`.  I expect that we'll wind up adding a lot of the same info as a payload on the `absl::Status` object, but the format may not be exactly the same.  @veblush took a stretch OKR this quarter to try to come up with a design for this.  None of our public APIs should be guaranteeing that this format or its contents will be a stable part of our API.

Another problem with `debug_error_string()` is that if we shove all the really useful info in there, then the normal error message field in the RPC status is basically useless.  I think that instead, we should be working to make the normal error message field contain all of the data that the user needs in the vast majority of cases, so that it becomes very rare for them to actually need to look at `debug_error_string()`.  This is consistent with how Java and Go handle their error messages (I don't think they even have an equivalent of `debug_error_string()`), and it was the intent behind #22883.

Reply:
I took a look at how we might add something along the lines of "happening during name resolution" to RPC error messages, but one issue I'm having here is that I can't see how to augment this message in a way that will reliably lead to something that is human-readable, while still preserving the information from the original error message.

When at how the code in https://github.com/grpc/grpc/blob/0a0aae8ee8f8dd1ea0d4a58ae8ac9700f1b82a4a/src/core/lib/transport/error_utils.cc#L48 determines the RPC error message from a `grpc_error*`, it looks complicated to pluck out the error message from the original `grpc_error`, and add "happened during name resolution" to it in a way that makes sense to a reader of the final message.

On the other hand, if we just dump the full error object as a JSON string, as `debug_error_string()` does, then IMO the result is actually more legible, simply because JSON is a nice way to organize disparate pieces of information which have been added by different subsystems (for example, the client channel's error message augmentation shouldn't stomp on an error message that was generated by the deadline filter or a resolver).

If there ends up being a way to add these different pieces of information to the resulting error message in a reliably legible way, then I'm open to use that. Otherwise, for `debug_error_string()` still seems like the most effective thing to leverage.